lib/rules/prefer-result-array-groups.ts

```diff
@@ -0,0 +1,104 @@
+    },
+    create(context) {
+        const strictTypes = context.options[0]?.strictTypes ?? true
+        const sourceCode = context.getSourceCode()
+
+        /**
+         * Create visitor
+         */
+        function createVisitor(
+            regexpContext: RegExpContext,
+        ): RegExpVisitor.Handlers {
+            const {
+                getAllCapturingGroups,
+                getCapturingGroupReferences,
+            } = regexpContext
+
+            const capturingGroups = getAllCapturingGroups()
+            if (!capturingGroups.length) {
+                return {}
+            }
+
+            for (const ref of getCapturingGroupReferences({ strictTypes })) {
+                if (
+                    ref.type === "ArrayRef" &&
+                    ref.kind === "index" &&
+                    ref.ref != null
+                ) {
+                    const cgNode = capturingGroups[ref.ref - 1]
+                    if (cgNode && cgNode.name) {
+                        const memberNode =
+                            ref.prop.type === "member" ? ref.prop.node : null
+                        context.report({
+                            node: ref.prop.node,
+                            messageId: "unexpected",
+                            fix:
+                                memberNode && memberNode.computed
+                                    ? (fixer) => {
+                                          const tokens = sourceCode.getTokensBetween(
+                                              memberNode.object,
+                                              memberNode.property,
+                                          )
+                                          let openingBracket = tokens.pop()
+                                          while (
+                                              openingBracket &&
+                                              !isOpeningBracketToken(
+                                                  openingBracket,
+                                              )
+                                          ) {
+                                              openingBracket = tokens.pop()
+                                          }
+                                          if (!openingBracket) {
+                                              // unknown ast
+                                              return null
+                                          }
+                                          return fixer.replaceTextRange(
+                                              [
+                                                  openingBracket.range![0],
+                                                  memberNode.range![1],
+                                              ],
+                                              `${
+                                                  memberNode.optional ? "" : "."
+                                              }groups.${cgNode.name}`,
```

Comment:
`groups` is [an optional property in TS](https://github.com/microsoft/TypeScript/blob/main/lib/lib.es2018.regexp.d.ts). This fix will cause a TS compiler error if strict null checks are enabled.

If the current file is a `.ts` file and strict null checks are enabled, then the fix should be `groups!.name`.

Does TS have an API to get the current TS config, so we can adjust the fix?

Reply:
Thank you for letting me know. I didn't consider for TypeScript.
I will change it to give `!` If strict type checking is enabled.

Reply:
I changed it to check the type before autofix.
Also, if a simple array type is specified, autofix will be stopped.

```ts
let match: any[] | null
while (match = regex.exec(foo)) {
    const p1 = match[1]
    // ...
}
```

Also, if it is a `any` type, `!` is not inserted.

Reply:
Amazing work! Thank you!